### ISSUE TITLE
broken exception handling

### ISSUE THREAD

**Issue by N-Coder** [fusepy/fusepy#97]
### Exceptions from init and destroy
According to the [libfuse docs](https://libfuse.github.io/doxygen/structfuse__operations.html#a3a1f9562a38871caaf4fc2de7ea5b470), the return value of the init function is not interpreted, but simply stored as private_data field of struct fuse_context without further looking at it. If the init function now throws a Python exception, the handler defined in [_wrapper](https://github.com/fusepy/fusepy/blob/master/fuse.py#L605) will consume the exception and simply return EFAULT. FUSE just stores that value, effectively suppressing the exception and continuing to mount a driver in a now possibly invalid state.
My recommendation would be to shut down the driver if `init` (or `destroy`) raises an exception and reraise the exception once the native driver terminated.

### handling SystemExit
Unfortunately, the exception handling will consume all exceptions no matter their nature, including `SystemExit`. This means that the driver can't be shutdown from any of the handlers, even using `sys.exit()`. This has also been mentioned in issue #21, although this problem is not related to threads or processes as mentioned in the comments to the issue, but to the broken exception handler of fusepy.

As python exceptions can't be propagated from within the python operations to the libfuse C code, I use the following [function](https://github.com/N-Coder/studip-fuse/blob/master/studip_fuse/__main__/fs_driver.py#L30) to shut down the driver in case of any problems:
```
def fuse_exit():
    from fuse import _libfuse, c_void_p

    fuse_ptr = c_void_p(_libfuse.fuse_get_context().contents.fuse)
    _libfuse.fuse_exit(fuse_ptr)

    # alternative without directly invoking native code
    # os.kill(os.getpid(), signal.SIGINT)
```
As mentioned in the other issue, if you don't want to call native code directly in that function, you could have the OS send SIGINT to your own process via `os.kill(os.getpid(), signal.SIGINT)`, invoking the libfuse SIGINT handler finally leading to the shutdown of the mount.
I'd vote for calling the native `fuse_exit` and making this function publicly available and thus fixing issue #21.

### OSErrors with invalid errnos
Some subclasses OSErrors (like [herror and gaierror](https://docs.python.org/3/library/socket.html#socket.herror)) use errnos for values that don't correspond to the normal OS errnos (but to host name resolution or getaddrinfo error code).
If such an error code is returned to libfuse, a warning about an invalid error code might be issued, or the file system operation might return an error code that makes no sense.
So those exception should be translated to standard errnos, e.g. EHOSTUNREACH, instead.

### Errors that are no subclass of OSError but can be mapped to an errno
There are some exceptions that are not a subclass of [OSError](http://www-numi.fnal.gov/offline_software/srt_public_context/WebDocs/Errors/unix_system_errors.html), but can directly mapped to an errno. Examples include:

- concurrent.futures.CancelledError and asyncio.CancelledError -> ECANCELED
- concurrent.futures.TimeoutError and asyncio.TimeoutError -> ETIMEDOUT
- requests.exceptions.ConnectionError -> ECONNRESET
- requests.exceptions.Timeout -> ETIMEDOUT
- aiohttp.ServerDisconnectedError -> ECONNRESET
- ...

As users might want to define handlers for further exceptions, the handler/mapper should probably be extensible.

### Summary
To sum up, is see the following problems:
- [ ] exceptions from `init` and `destroy` should not be consumed, but lead to termination of the driver
- [ ] `sys.exit()` will be consumed by the current exception handler, add function for shutting down the driver
- [ ] some OSErrors (e.g. herror and gaierror) have negative errnos, which should be translated to standard errnos
- [ ] some python errors (e.g. CancelledError) have no errnos, but could be mapped to one

My attempt at fixing these issues can be found [here](https://github.com/N-Coder/studip-fuse/blob/master/studip_fuse/__main__/fs_driver.py#L40). If you like my solution, I'd be glad to make a pull request out if it.

**Comment by terencehonles** [comment]
Thanks for the report. As you described the issue I was predicting what your solutions would be (and they are what I would do). I picked up this project from the original author and haven't had as much time as I would have liked to maintain it, but it is unsurprising this issue exists.

We are currently working towards a major release, but it doesn't seem like these changes are backwards incompatible.

- If you don't mind can you put the init and destroy fix in a PR with handling SystemExit (use the fuse_exit please). This PR should be pretty straight forward and not need much discussion.
- The error handling code you mentioned should probably be in a separate PR because it may need more back and forth
- finally the `_wrapper` function writing to sys.stderr I want to look at a little more, but I think regardless it should log a message because that can be configured by the consumer of the library

**Comment by N-Coder** [comment]
I created PR #103 for gracefully handling critical exceptions from init or such as SystemExit, please see the commit message for details.
This should address points 1 and 2 in my ToDo list above.
I didn't include the handling of more specific exceptions (as mentioned in points 3 and 4 in my ToDo list and the second point of @terencehonles list) to leave the exact approach for each Exception open for debate. 
Nonetheless I did include a small fix for the invalid errnos in OSError and also dropped `traceback.print_exc` in favor of `logging.error`.

**Comment by eestrada** [comment]
@N-Coder It is good to know what the actual issue is for #21. Thanks for giving such a complete description of what the core problem actually is.

**Comment by koldinger** [comment]
I've run into a similar problem to the invalid errno issue.   I'm wrapping a fusepy filesystem around http requests going through the Requests library.   If Requests gets an error, it throws a requests.exceptinos.HTTPError exception, which extends OSError, but has None as it's errno.   This causes the "e.errorno > 0" test to fail, and the whole thing to punt out.

**Comment by N-Coder** [comment]
@rianhunter already made a PR with a fix for this:
https://github.com/fusepy/fusepy/pull/125/files#diff-1d782d08fb7c4569538882342ebbc21cR818
Unfortunately, it is mixed with support for OpenBSD and changes to how the exception management should work.

I created a PR fixing this issue in @s-m-e's fork (which I'm going to use in the future): https://github.com/pleiszenburg/refuse/pull/24
